engine/src/main/java/com/google/android/fhir/sync/upload/Uploader.kt

```diff
@@ -38,67 +43,131 @@ import timber.log.Timber
+        bundleEntry.hasResponse() ->
+          BundleComponentUploadResponse(localChanges, bundleEntry.response)
+        else ->
+          throw IllegalStateException(
+            "Unknown response: $bundleEntry for Bundle Request at index $index",
+          )
+      }
+    }
   }
 
-  private suspend fun handleUploadRequest(uploadRequest: UploadRequest): UploadRequestResult {
+  private suspend fun handleUploadRequest(
+    mappedUploadRequest: UploadRequestGeneratorOutput,
+  ): UploadRequestResult {
     return try {
-      val response = dataSource.upload(uploadRequest)
+      val response = dataSource.upload(mappedUploadRequest.generatedRequest)
       when {
-        response is Bundle && response.type == Bundle.BundleType.TRANSACTIONRESPONSE ->
-          UploadRequestResult.Success(response)
         response is OperationOutcome && response.issue.isNotEmpty() ->
           UploadRequestResult.Failure(
             ResourceSyncException(
-              uploadRequest.resource.resourceType,
+              mappedUploadRequest.generatedRequest.resource.resourceType,
               FHIRException(response.issueFirstRep.diagnostics),
             ),
           )
+        (response is DomainResource || response is Bundle) &&
+          (response !is IBaseOperationOutcome) ->
+          handleUploadResponse(mappedUploadRequest, response)
         else ->
           UploadRequestResult.Failure(
             ResourceSyncException(
-              uploadRequest.resource.resourceType,
-              FHIRException("Unknown response for ${uploadRequest.resource.resourceType}"),
+              mappedUploadRequest.generatedRequest.resource.resourceType,
+              FHIRException(
+                "Unknown response for ${mappedUploadRequest.generatedRequest.resource.resourceType}",
+              ),
             ),
           )
       }
     } catch (e: Exception) {
       Timber.e(e)
-      UploadRequestResult.Failure(ResourceSyncException(ResourceType.Bundle, e))
+      UploadRequestResult.Failure(
+        ResourceSyncException(mappedUploadRequest.generatedRequest.resource.resourceType, e),
+      )
     }
   }
 
   private sealed class UploadRequestResult {
-    data class Success(val resource: Resource) : UploadRequestResult()
+    data class Success(
+      val successfulUploadRequestResponses: List<SuccessfulUploadRequestResponse>,
+    ) : UploadRequestResult()
 
     data class Failure(val exception: ResourceSyncException) : UploadRequestResult()
   }
 }
 
 sealed class UploadSyncResult {
   data class Success(
-    val localChanges: List<LocalChange>,
-    val responseResources: List<Resource>,
+    val uploadResponses: List<SuccessfulUploadRequestResponse>,
   ) : UploadSyncResult()
 
   data class Failure(val syncError: ResourceSyncException, val localChangeToken: LocalChangeToken) :
     UploadSyncResult()
 }
+
+sealed class SuccessfulUploadRequestResponse(
```

Comment:
rename to be more accurate. here and inherited classes below
```suggestion
sealed class SuccessfulUploadResponseMapping(
```

Reply:
Done